---
title: "S2 MET GWAS Analysis"
output: html_notebook
bibliography: C:/Users/Jeff/Documents/Literature/library.bib
---

This notebook will outline the analysis of GWAS mapping procedures using the S2MET data.


## Introduction

Load packages and set directories

```{r setup}

library(tidyverse)

# Project and other directories
proj_dir <- "C:/Users/Jeff/Google Drive/Barley Lab/Projects/S2MET_Mapping//"
alt_proj_dir <- "C:/Users/Jeff/Google Drive/Barley Lab/Projects/S2MET"

# Geno, pheno, and enviro data
geno_dir <-  "C:/Users/Jeff/Google Drive/Barley Lab/Projects/Genomic Selection/Genotypic Data/GBS Genotype Data/"
pheno_dir <- file.path(alt_proj_dir, "Phenotype_Data/")
env_var_dir <- file.path(alt_proj_dir, "Environmental_Variables")

# Other directories
fig_dir <- file.path(proj_dir, "Figures/")
map_dir <- file.path(proj_dir, "Mapping")
entry_dir <- file.path(alt_proj_dir, "Plant_Materials")
analysis_dir <- file.path(proj_dir, "Analysis")
result_dir <- file.path(proj_dir, "Results")

```


## Methods

The first mapping trial used environments as fixed effects and fit the following model:

$$
\mathbf{y = X \beta + Zg + Sa + e}
$$

where $\mathbf{y}$ is the response vector, $\mathbf{X}$ is the fixed effect incidence matrix modeling environment as a fixed effect, $\mathbf{\beta}$ is the vector of fixed environmental effects, $\mathbf{Z}$ is the incidence matrix modeling the random genetic background effect, $\mathbf{g}$ is the vector of random genetic background effects, $\mathbf{S}$ is the SNP marker matrix, and $\mathbf{a}$ is the vector of fixed additive SNP effects. $\mathbf{g}$ is assumed to be distributed such that $\mathbf{g} \sim N(0, \mathbf{K} \sigma^2_g )$, where $\mathbf{K}$ models the population structure among genotypes.

I used the `GWAS()` function in the `rrBLUP` package to fit this model

```{r main.model1}

# Load the data
load(file.path(map_dir, "S2MET_gwas_main_results.RData"))

# Gather data
gwas_main_out1 <- gwas_main_out %>% 
  gather(trait, neg_log_p, -marker:-pos) %>%
  # Convert -log_p to p
  mutate(p_value = exp(-1 * neg_log_p))

# Calculate the significant p_value at a FDR of x
fdr = 0.05

gwas_main_out2 <- gwas_main_out1 %>%
  group_by(trait) %>%
  mutate(neg_log_fdr = fdr2(p = neg_log_p, fdr.level = fdr)$fdr.10) %>%
  ungroup()


# Q-Q Plot per trait
gwas_main_out2 %>%
  ggplot(aes(sample = p_value)) +
  geom_point(stat = "qq") +
  facet_wrap(~ trait, ncol = 2)

```


```{r plot.gwas}
# Plot the GWAS results
gwas_main_out2 %>%
  mutate(chrom = as.factor(chrom)) %>%
  ggplot(aes(x = pos, y = neg_log_p, col = chrom)) +
  geom_point() +
  facet_grid(trait ~ chrom, space = "free_x", scale = "free") +
  geom_hline(aes(yintercept = neg_log_fdr)) +
  scale_color_discrete(guide = FALSE) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))



```










