---
title: "S2 MET GWAS Analysis"
output: html_notebook
bibliography: C:/Users/Jeff/Documents/Literature/library.bib
---

This notebook will outline the analysis of GWAS mapping procedures using the S2MET data.


## Introduction

Load packages and set directories

```{r setup}

library(tidyverse)

# Project and other directories
source("C:/Users/Jeff/Google Drive/Barley Lab/Projects/S2MET_Mapping/source.R")

```


## Methods

### Population Structure

First examine population structure within the germplasm

```{r pop.str}

# Calculate the A matrix
A <- A.mat(X = s2_imputed_mat_use, min.MAF = 0, max.missing = 1)

## Visualize pop structure
A_prcomp <- prcomp(A)

# Extract the proportion explained by each PC
lambda <- round(A_prcomp$sdev / sum(A_prcomp$sdev), 3) * 100

# Convert to data.frame
A_prcomp_df <- A_prcomp$rotation %>% 
  as.data.frame() %>% 
  rownames_to_column("line_name")

A_prcomp_df1 <- left_join(A_prcomp_df, entry_list, c("line_name" = "Line")) %>% 
  dplyr::select(line_name, program = Program, parent = `Parent?`, starts_with("PC")) %>%
  mutate(program = ifelse(str_detect(line_name, "^2MS"), "M2", program))

# Plot
(g_pop_str <- A_prcomp_df1 %>% 
  ggplot(aes(x = PC1, y = PC2, col = program)) + 
  geom_point(size = 2) + 
  ylab(str_c("PC2 (", lambda[2], "%)")) +
  xlab(str_c("PC1 (", lambda[1], "%)")) +
  scale_color_discrete(guide = guide_legend(title = "Breeding\nProgram")) +
  labs(
    title = "Population Structure of the S2MET Genotypes",
    subtitle = "The 'M2' program includes genotypes derived\nfrom crosses made among genotypes from the other programs."
  ) +
  theme_bw())

# Save this
save_file <- file.path(fig_dir, "population_structure.jpg")
ggsave(filename = save_file, plot = g_pop_str, width = 6, height = 5)

```


### GWAS

#### Stability QTL

First we will map QTL for trait stability. The stability coefficients are first calculated using FW regression, and the coefficients are then used in the GWAS mixed model:

$$
y_{i} = \mu + \sum^m_{m=1} x_{im} \alpha_m + g_{i} + \epsilon_{i}
$$

which models the stability of genotype *i* as the sum of the grand mean, the sum of the marker allele effects carried by genotype *i*, the residual genetic effect of genotype *i*, and the residual effect associated with the observation.

In vector notation this becomes:


$$
\mathbf{y = X \beta + Wm + Zu + e}
$$

where $\mathbf{y}$ is the response vector, $\mathbf{X}$ is the fixed effect incidence matrix modeling environment as a fixed effect, $\mathbf{\beta}$ is the vector of fixed environmental effects, $\mathbf{Z}$ is the incidence matrix modeling the random genetic background effect, $\mathbf{g}$ is the vector of random genetic background effects, $\mathbf{W}$ is the SNP marker matrix, and $\mathbf{m}$ is the vector of fixed additive SNP effects. $\mathbf{g}$ is assumed to be distributed such that $\mathbf{g} \sim N(0, \mathbf{K} \sigma^2_g )$, where $\mathbf{K}$ models the population structure among genotypes.

I used the `GWAS()` function in the `rrBLUP` package to fit this model. The first round used the P + K model, and the second round used the K only model.


#### Main Effect QTL and QTLxE

The above model does not account for QTLxE interaction. We will therefore split up the modeling of main effect QTL, QTLxE, and stability QTL.

We fit the following model for QTL main effects and QTLxE

$$
y_{ij} = \mu + E_j + \sum^m_{m=1} x_{im} (\alpha_m + \delta_{mj}) + g_{ij} + \epsilon_{ij}
$$

which models the mean of genotype *i* in environment *j* as the sum of the grand mean, the effect of the *j*th environment, the sum of the marker allele effects ($\alpha_m$) and environment-specific marker deviations ($\delta_{mj}$) carried by genotype *i*, the residual genetic effect of genotype *i* in environment *j*, and the residual effect associated with the observation.

In vector notation this becomes:


$$
\mathbf{y = X \beta + Wm + Vd + Zu + e}
$$

where $\mathbf{y}$ is the response vector, $\mathbf{X}$ is the fixed effect incidence matrix modeling environment as a fixed effect, $\mathbf{\beta}$ is the vector of fixed environmental effects, $\mathbf{Z}$ is the incidence matrix modeling the random genetic background effect, $\mathbf{g}$ is the vector of random genetic background effects, $\mathbf{W}$ is the SNP marker matrix, $\mathbf{m}$ is the vector of fixed additive SNP effects, $\mathbf{V}$ is the incidence matrix of SNPs across environments, and $\mathbf{d}$ is the vector of environment-specific SNP deviations. $\mathbf{g}$ is assumed to be distributed such that $\mathbf{g} \sim N(0, \mathbf{K} \sigma^2_g )$, where $\mathbf{K}$ models the population structure among genotypes.


## Results


```{r main.model1}

# Load the data
load(file.path(result_dir, "S2MET_gwas_fw_results.RData"))

# Gather data
gwas_main_out1 <- gwas_fw_out %>% 
  gather(trait, neg_log_p, -marker:-pos) %>%
  # Convert -log_p to p
  mutate(p_value = 10 ^ (-1 * neg_log_p))

# Correct the p_values for an FDR of 0.05
gwas_main_out2 <- gwas_main_out1 %>%
  group_by(trait) %>%
  mutate(p_adj = p.adjust(p = p_value, method = "fdr"),
         neg_log_p_adj = -log10(p_adj),
         neg_log_fdr05 = -log10(0.05),
         neg_log_fdr10 = -log10(0.10)) %>%
  ungroup() %>%
  separate(trait, c("trait", "coef"), sep = "_") %>%
  mutate(coef = ifelse(coef == "b", "Sensitivity", "GenotypeMean"))


# Q-Q Plot per trait
gwas_qq <- gwas_main_out2 %>%
  group_by(trait, coef) %>%
  arrange(trait, coef, p_value) %>% 
  mutate(exp_neg_log_p = seq_along(neg_log_p), 
         exp_neg_log_p = -(log10(exp_neg_log_p / (length(exp_neg_log_p)+1)))) %>%
  arrange(trait, coef, chrom, pos)

(g_gwas_qq <- gwas_qq %>%
  ggplot(aes(x = exp_neg_log_p, y = neg_log_p)) +
  geom_point() +
  geom_abline(intercept = 0, slope = 1, col = "red", lwd = 1) +
  facet_grid(trait ~ coef) +
  ylab("Observed -log10(p)") +
  xlab("Expected -log10(p)") +
  labs(
    title = "Q-Q Plot of Association Results"
  ) +
  theme_bw() )

# Save this
save_file <- file.path(fig_dir, "gwas_qq.jpg")
ggsave(filename = save_file, plot = g_gwas_qq, width = 6, height = 6)


```


```{r plot.gwas}

# Plot the GWAS results
(g_fw_gwas <- gwas_main_out2 %>%
  mutate(chrom = as.factor(chrom),
         pos = pos / 1000000) %>%
  ggplot(aes(x = pos, y = neg_log_p_adj, col = chrom)) +
  geom_point() +
  geom_hline(aes(yintercept = neg_log_fdr05, lty = "FDR 05%")) +
  geom_hline(aes(yintercept = neg_log_fdr10, lty = "FDR 10%")) +
  facet_grid(trait + coef ~ chrom, space = "free_x", scale = "free", switch = "x") +
  ylab("Adjusted -log(p)") +
  xlab("Position (Mbp)") +
  scale_color_brewer(palette = "Set2", drop = FALSE, guide = FALSE) +
  scale_linetype(guide = guide_legend(title = NULL)) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    panel.border = element_blank()
  ) +
  labs(
    title = "GWAS of Genotypic Main Effects and Sensitivity Coefficients"
  ))


# Save this
save_file <- file.path(fig_dir, "gwas_fw_results.jpg")
ggsave(filename = save_file, plot = g_fw_gwas, width = 12, height = 7, dpi = 500)


```










