---
title: "S2 MET GWAS Analysis"
output: html_notebook
bibliography: C:/Users/Jeff/Documents/Literature/library.bib
---

This notebook will outline the analysis of GWAS mapping procedures using the S2MET data.


## Introduction

Load packages and set directories

```{r setup}

library(tidyverse)

# Project and other directories
source("C:/Users/Jeff/Google Drive/Barley Lab/Projects/S2MET_Mapping/source.R")

```


## Methods

First examine population structure within the material

```{r pop.str}

# Calculate the A matrix
A <- A.mat(X = s2_imputed_mat_use, min.MAF = 0, max.missing = 1)

## Visualize pop structure
A_prcomp <- prcomp(A)

# Extract the proportion explained by each PC
lambda <- round(A_prcomp$sdev / sum(A_prcomp$sdev), 3) * 100

# Convert to data.frame
A_prcomp_df <- A_prcomp$rotation %>% 
  as.data.frame() %>% 
  rownames_to_column("line_name")

A_prcomp_df1 <- left_join(A_prcomp_df, entry_list, c("line_name" = "Line")) %>% 
  dplyr::select(line_name, program = Program, parent = `Parent?`, starts_with("PC")) %>%
  mutate(program = ifelse(str_detect(line_name, "^2MS"), "M2", program))

# Plot
(g_pop_str <- A_prcomp_df1 %>% 
  ggplot(aes(x = PC1, y = PC2, col = program)) + 
  geom_point(size = 2) + 
  ylab(str_c("PC2 (", lambda[2], "%)")) +
  xlab(str_c("PC1 (", lambda[1], "%)")) +
  scale_color_discrete(guide = guide_legend(title = "Breeding\nProgram")) +
  labs(
    title = "Population Structure of the S2MET Genotypes",
    subtitle = "The 'M2' program includes genotypes derived\nfrom crosses made among genotypes from the other programs."
  ) +
  theme_bw())

# Save this
save_file <- file.path(fig_dir, "population_structure.jpg")
ggsave(filename = save_file, plot = g_pop_str, width = 6, height = 5)

```



The first mapping trial used environments as fixed effects and fit the following model:

$$
\mathbf{y = X \beta + Zg + Sa + e}
$$

where $\mathbf{y}$ is the response vector, $\mathbf{X}$ is the fixed effect incidence matrix modeling environment as a fixed effect, $\mathbf{\beta}$ is the vector of fixed environmental effects, $\mathbf{Z}$ is the incidence matrix modeling the random genetic background effect, $\mathbf{g}$ is the vector of random genetic background effects, $\mathbf{S}$ is the SNP marker matrix, and $\mathbf{a}$ is the vector of fixed additive SNP effects. $\mathbf{g}$ is assumed to be distributed such that $\mathbf{g} \sim N(0, \mathbf{K} \sigma^2_g )$, where $\mathbf{K}$ models the population structure among genotypes.

I used the `GWAS()` function in the `rrBLUP` package to fit this model

```{r main.model1}

# Load the data
load(file.path(map_dir, "S2MET_gwas_fw_results.RData"))

# Gather data
gwas_main_out1 <- gwas_fw_out %>% 
  gather(trait, neg_log_p, -marker:-pos) %>%
  # Convert -log_p to p
  mutate(p_value = exp(-1 * neg_log_p))

# Calculate the significant p_value at a FDR of x
fdr = 0.05

gwas_main_out2 <- gwas_main_out1 %>%
  group_by(trait) %>%
  mutate(neg_log_fdr = sommer::fdr2(p = neg_log_p, fdr.level = fdr)$fdr.10) %>%
  ungroup() %>%
  separate(trait, c("trait", "coef"), sep = "_")


# Q-Q Plot per trait
gwas_main_out2 %>%
  ggplot(aes(sample = p_value)) +
  geom_point(stat = "qq") +
  facet_wrap(~ trait, ncol = 2)

```


```{r plot.gwas}
# Plot the GWAS results
(g_fw_gwas <- gwas_main_out2 %>%
  mutate(chrom = as.factor(chrom),
         pos = pos / 1000000) %>%
  ggplot(aes(x = pos, y = neg_log_p, col = chrom)) +
  geom_point() +
  geom_hline(aes(yintercept = neg_log_fdr)) +
  facet_grid(trait + coef ~ chrom, space = "free_x", scale = "free", switch = "x") +
  ylab("-log(p)") +
  xlab("Position (Mbp)") +
  scale_color_discrete(guide = FALSE) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    panel.border = element_blank()
  ) +
  labs(
    title = "GWAS of Genotypic Main Effects and Sensitivity Coefficients",
    caption = "Horizontal line: significance level at FDR = 0.05"
  ))

# Save this
save_file <- file.path(fig_dir, "gwas_fw_results.jpg")
ggsave(filename = save_file, plot = g_fw_gwas, width = 10, height = 8)


```










