---
title: "S2 MET GWAS Analysis"
output: html_notebook
bibliography: C:/Users/Jeff/Documents/Literature/library.bib
---

This notebook will outline the analysis of GWAS mapping procedures using the S2MET data.

The GWAS will examine main effect QTL, QTLxE, and QTL influencing stability.


## Introduction

Load packages and set directories

```{r setup}

library(tidyverse)
library(purrrlyr)
library(readxl)
library(lme4)
library(rrBLUP)
library(broom)
library(GenomicRanges)
library(qvalue)
library(ggforce)
# library(hierGWAS)

# Project and other directories
source("C:/Users/Jeff/Google Drive/Barley Lab/Projects/S2MET_Mapping/source.R")

tp_geno <- tp_geno_multi

load(file.path(geno_dir, "S2_genos_hmp.RData"))

# Filter environments for those in which the TP was observed
S2_MET_BLUEs_use <- S2_MET_BLUEs %>% 
  filter(line_name %in% tp_geno)

# Filter the genotype matrix for the TP only
# M <- s2_imputed_mat[tp_geno,]
M <- S2TP_imputed_multi_genos_mat[tp_geno,]

```






## Methods

Convert the markers into a GRanges object, then align with the gene annotations

```{r marker.annotation}

# Load the genome annotation
ann_dir <- "C:/Users/Jeff/Google Drive/Barley Lab/Projects/Genomics/Annotation"
load(file.path(ann_dir, "barley_genomic_ranges.RData"))


# Convert to GRanges object
snp_info <- S2TP_imputed_multi_genos_hmp %>% 
  select(rs:cM_pos) %>%
  separate(alleles, c("ref", "alt"), "/") %>%
  mutate(chrom = str_c("chr", chrom, "H"))
  

snp_info_grange <- GRanges(seqnames = snp_info$chrom,
                           ranges = IRanges(snp_info$pos, snp_info$pos),
                           ref = snp_info$ref, alt = snp_info$alt, 
                           cM_pos = snp_info$cM_pos, rs = snp_info$rs)

## Align the markers to the barley granges information
snp_info_grange_overlap <- barley_grange_list %>% 
  as.list() %>% 
  map(~ mergeByOverlaps(query = snp_info_grange, subject = .))

## Save
save_file <- file.path(result_dir, "snp_genome_annotation.RData")
save("snp_info_grange_overlap", file = save_file)

```




## Results

### Population Structure

First examine population structure within the germplasm

```{r pop.str}

# Calculate the A matrix
# A_all <- A.mat(X = M, min.MAF = 0, max.missing = 1)
A_tp <- A.mat(X = M, min.MAF = 0, max.missing = 1)

A_list <- list(tp = A_tp)

## Visualize pop structure
A_prcomp <- map(A_list, prcomp)

# Extract the proportion explained by each PC
lambda <- map(A_prcomp, "sdev") %>% map(~round(. / sum(.), 3) * 100)

# Convert to data.frame
A_prcomp_df_list <- map(A_prcomp, "rotation") %>% 
  map(~as.data.frame(.) %>% rownames_to_column("line_name"))

A_prcomp_df <- A_prcomp_df_list %>%
  map(~left_join(., entry_list, by = c("line_name" = "Line")) %>%
        select(line_name, program = Program, parent = `Parent?`, starts_with("PC")) %>%
        mutate(program = if_else(str_detect(line_name, "^2MS"), "M2", program))) %>%
  list(., c("All", "TP")) %>% 
  pmap(~mutate(.x, population = .y)) %>% 
  bind_rows()

A_prcomp_df <- A_prcomp_df %>%
  filter(population == "TP")


# # Plot
# (g_pop_str <- A_prcomp_df %>%
#   ggplot(aes(x = PC1, y = PC2, col = program)) +
#   geom_point(size = 2) +
#   ylab("PC2") +
#   xlab("PC1") +
#   scale_color_discrete(guide = guide_legend(title = "Breeding\nProgram")) +
#   facet_grid(~ population) +
#   labs(
#     title = "Population Structure of the S2MET Genotypes",
#     subtitle = "The 'M2' program includes genotypes derived\nfrom crosses made among genotypes from the other programs."
#   ) +
#   theme_bw())
# 
# # Save this
# save_file <- file.path(fig_dir, "population_structure.jpg")
# ggsave(filename = save_file, plot = g_pop_str, width = 7, height = 5)


# Plot
(g_pop_str <- A_prcomp_df %>%
    filter(population == "TP") %>%
  ggplot(aes(x = PC1, y = PC2, col = program)) +
  geom_point(size = 2) +
  ylab("PC2") +
  xlab("PC1") +
  scale_color_discrete(guide = guide_legend(title = "Breeding\nProgram")) +
  facet_grid(~ population) +
  labs(title = "Population Structure of the S2MET Genotypes",
       caption = str_c("No. markers = ", ncol(M))) +
  theme_bw())

# Save this
save_file <- file.path(fig_dir, "population_structure.jpg")
ggsave(filename = save_file, plot = g_pop_str, width = 5, height = 5)

```


How is population structure correlated with the traits?

First, fit a model to calculate BLUEs for each genotype across all environments

```{r mme.blues, eval=FALSE}

trait_BLUEs <- S2_MET_BLUEs_use %>%
  group_by(trait) %>%
  do({
    df <- .
    # Standard errors as weights
    wts <- df$std_error^2
    
    # Control the model
    lmer_control <- lmerControl(check.nobs.vs.nlev = "ignore", check.nobs.vs.nRE = "ignore")
    
    # Set the contrasts of line_name and environment
    df_use <- df %>%
      mutate_at(vars(line_name, environment), as.factor)
    
    # Fit the model
    # fit <- lmer(value ~ -1 + line_name + environment + (1|line_name:environment), data = df_use,
    #             weights = wts, control = lmer_control)
    
    fit <- lmer(value ~ -1 + line_name + environment + (1|line_name:environment), data = df_use,
                control = lmer_control)
    
    # Extract BLUEs and return
    tidy(fit) %>% 
      filter(str_detect(term, "line_name"), group == "fixed") %>% 
      mutate(term = str_replace(term, "line_name", "")) %>% 
      select(line_name = term, value = estimate) })

## Correlation among the BLUEs
trait_BLUEs %>% 
  spread(trait, value) %>% 
  .[,-1] %>% 
  cor()

# Combine the trait and PCs, duplicating the BLUEs to make summarizing smoother
trait_pop_str <- A_prcomp_df %>% 
  select(line_name:parent, PC1:PC3) %>% 
  gather(PC, eigenvalue, PC1:PC3) %>% 
  left_join(., trait_BLUEs, by = "line_name")

# Correlate trait BLUEs with eigenvectors
trait_pop_str_summ <- trait_pop_str %>% 
  group_by(PC, trait) %>% 
  summarize(trait_str_cor = cor(eigenvalue, value))


# Plot
(g_trait_pop_str <- trait_pop_str %>% 
  ggplot(aes(x = eigenvalue, y = value)) +
  geom_smooth(method = "lm", se = FALSE) +
  geom_point(aes(col = program)) +
  facet_grid(trait ~ PC, scales = "free") +
  ylab("Genotypic Value") +
  xlab("Eigenvalue") +
  labs(
    title = "Relationship Between Trait and Population Structure"
  ))

save_file <- file.path(fig_dir, "trait_pop_str_relationship.jpg")
ggsave(filename = save_file, plot = g_trait_pop_str, width = 6, height = 8, dpi = 500)

```




### GWAS

The baseline model for GWAS is as follows:

$$
\mathbf{y = X \beta + Wm + Zu + e},
$$

where $\mathbf{y}$ is the response vector, $\mathbf{X}$ is the fixed effect incidence matrix modeling environment as a fixed effect, $\mathbf{\beta}$ is the vector of fixed environmental effects, $\mathbf{Z}$ is the incidence matrix modeling the random genetic background effect, $\mathbf{g}$ is the vector of random genetic background effects, $\mathbf{W}$ is the SNP marker matrix, and $\mathbf{m}$ is the vector of fixed additive SNP effects. $\mathbf{g}$ is assumed to be distributed such that $\mathbf{g} \sim N(0, \mathbf{K} \sigma^2_g )$, where $\mathbf{K}$ models the population structure among genotypes.

I use population parameters previously determined [P3D; @Kang2010] to speed up the models without reducing detection power significantly. I also the `gwas()` function in my `pbr` package. Models included K and G [@Bernardo2013b; @Chen2016].


Load the results and adjust the p-values

```{r load.and.adj}

## Main effect QTL
load(file.path(result_dir, "S2MET_gwas_genotype_mean.RData"))

# Convert to DF and filter for main effect assocations
gwas_geno_mean <- gwas_main %>% 
  map_df(~mutate(.$scores, model = .$metadata$model)) %>%
  filter(term == "main_effect")

## Adjust the p-values
gwas_geno_mean_adj <- gwas_geno_mean %>%
  group_by(., trait, model) %>% 
  mutate(p_adj = p.adjust(p_value, method = "fdr"),
         q_value = qvalue(p = p_value)$qvalue,
         local_fdr = qvalue(p = p_value)$lfdr,
         neg_log10_fdr05 = -log10(0.05),
         neg_log10_fdr10 = -log10(0.10)) %>%
  mutate_at(vars(p_value, p_adj, q_value), funs(neg_log10 = -log10(.))) %>%
  ungroup()


# Load the data
# load(file.path(result_dir, "S2MET_pheno_fw_gwas_results.RData"))
load(file.path(result_dir, "S2MET_pheno_fw_gwas_multi_geno_results.RData"))

# Load the permutation results
load(file.path(result_dir, "S2MET_pheno_fw_gwas_permutation.RData"))

# Add the model name to the df
gwas_pheno_fw1 <- gwas_pheno_fw %>% 
  map_df(., ~mutate(.$scores, model = .$metadata$model) %>%
           filter(term == "main_effect")) %>%
  as_data_frame()

## Adjust the p-values
gwas_pheno_fw_adj <- gwas_pheno_fw1 %>%
  group_by(., trait, model) %>% 
  mutate(p_adj = p.adjust(p_value, method = "fdr"),
         q_value = qvalue(p = p_value)$qvalue,
         local_fdr = qvalue(p = p_value)$lfdr,
         neg_log10_fdr05 = -log10(0.05),
         neg_log10_fdr10 = -log10(0.10)) %>%
  mutate_at(vars(p_value, p_adj, q_value), funs(neg_log10 = -log10(.)))


# Reformat the permutation results
gwas_pheno_fw_perm <- pheno_fw_permute_out %>% 
  map_df(bind_rows) %>% 
  as_data_frame()

# Combine the realized GWAS results with the permutation results
gwas_pheno_fw_perm_ci <- gwas_pheno_fw1 %>%
  unnest(estimate) %>% 
  filter(model == "G") %>% 
  select(marker, chrom, pos, trait, term, realized_W = W_statistic, realized_estimate = estimate) %>%
  full_join(., gwas_pheno_fw_perm)

# Calculate the p-value by determining the probability of observing test statistics
# that are greater than the observed test statistic under H_0 = TRUE
gwas_pheno_fw_perm_pval <- gwas_pheno_fw_perm_ci %>% 
  mutate(beta_contradictory = abs(estimate) >= abs(realized_estimate),
         W_contradictory = abs(W_statistic) >= abs(realized_W)) %>% 
  group_by(trait, marker) %>% 
  mutate(p_value_beta = mean(beta_contradictory), 
         p_value_W = mean(W_contradictory),
         n_perm = n()) %>%
  select(marker:trait, realized_estimate, p_value_beta, p_value_W, n_perm) %>%
  distinct()

# Load the marker effect stability test results
load(file.path(result_dir, "S2MET_pheno_mar_eff_fw_results.RData"))

# Load the genome annotation intersections
load(file.path(result_dir, "snp_genome_annotation.RData"))


```



#### Main Effect QTL

This model was used to detect associations between loci and the genotypic mean of each trait.

Load the data and plot results

```{r main.effect.qtl}

## Quality control
## QQ plot
qq_gwas_geno_mean_adj <- gwas_geno_mean_adj %>% 
  arrange(trait, model, p_value) %>% 
  mutate(p_exp = punif(q = seq(0, 1, length.out = n())), 
         neg_log10_p_exp = -log10(p_exp))

# Plot
g_qq <- qq_gwas_geno_mean_adj %>%
  ggplot(aes(x = neg_log10_p_exp, y = p_value_neg_log10, col = model)) + 
  geom_point() + 
  geom_abline(slope = 1, intercept = 0) + 
  facet_wrap(~ trait, ncol = 2) +
  labs(title = "QQ Plot") +
  theme_bw()

save_file <- file.path(fig_dir, "gwas_pheno_genotype_mean_qq.jpg")
ggsave(filename = save_file, plot = g_qq, width = 5, height = 7)


## Common plot layers for the manhattan plots
g_add <- list(geom_point(),
              geom_hline(aes(yintercept = neg_log10_fdr05, lty = "FDR 05%")),
              geom_hline(aes(yintercept = neg_log10_fdr10, lty = "FDR 10%")),
              scale_color_manual(values = set_names(rep(c("black", "grey75"), length.out = 7), seq(1, 7)),
                                 guide = FALSE),
              ylab("-log(q)"),
              xlab("Position"),
              labs(title = "Genomwide Assocation Analysis for Genotype Mean"),
              theme_bw(),
              theme(panel.spacing.x = unit(x = 0, units = "in"),
                    panel.border = element_blank(),
                    axis.text.x = element_text(angle = 45, hjust = 1),
                    legend.position = "bottom"))

## Manhattan plot
g_man <- gwas_geno_mean_adj %>%
  mutate(chrom = as.factor(chrom)) %>% 
  # ggplot(aes(x = pos / 1000000, y = neg_log10_p_adj, group = chrom, col = chrom)) + 
  ggplot(aes(x = pos / 1000000, y = q_value_neg_log10, group = chrom, col = chrom)) +
  facet_grid(trait + model ~ chrom, switch = "x", scales = "free", space = "free_x") +
  g_add

save_file <- file.path(fig_dir, "gwas_pheno_genotype_mean_manhattan.jpg")
ggsave(filename = save_file, plot = g_man, width = 9, height = 12)


## Only consider the G models
g_man_Gmodels <- gwas_geno_mean_adj %>%
  filter(model %in% c("G", "QG")) %>%
  mutate(chrom = as.factor(chrom)) %>%
  # ggplot(aes(x = pos / 1000000, y = neg_log10_p_adj, group = chrom, col = chrom)) + 
  ggplot(aes(x = pos / 1000000, y = q_value_neg_log10, group = chrom, col = chrom)) + 
  facet_grid(trait + model ~ chrom, switch = "x", scales = "free", space = "free_x") +
  g_add

save_file <- file.path(fig_dir, "gwas_pheno_genotype_mean_manhattan_Gmodels.jpg")
ggsave(filename = save_file, plot = g_man_Gmodels, width = 9, height = 10)


## Just the G model
g_man_Gmodel <- gwas_geno_mean_adj %>%
  filter(model == "G") %>%
  mutate(chrom = as.factor(chrom)) %>%
  # ggplot(aes(x = pos / 1000000, y = neg_log10_p_adj, group = chrom, col = chrom)) + 
  ggplot(aes(x = pos / 1000000, y = q_value_neg_log10, group = chrom, col = chrom)) + 
  facet_grid(trait ~ chrom, switch = "x", scales = "free", space = "free_x") +
  g_add +
  labs(subtitle = "G Model Only")

save_file <- file.path(fig_dir, "gwas_pheno_genotype_mean_manhattan_Gmodel.jpg")
ggsave(filename = save_file, plot = g_man_Gmodel, width = 9, height = 8)



```



#### Stability QTL

First we will map QTL for trait stability. The stability coefficients are first calculated using FW regression, and the coefficients are then used in the GWAS mixed model:

$$
\mathbf{y = 1\beta + Wm + Zu + e}
$$

where $\mathbf{y}$ is the response vector, $\mathbf{1}$ is a vector of 1s for the grand mean, $\mathbf{\beta}$ is the grand mean, $\mathbf{Z}$ is the incidence matrix modeling the random genetic background effect, $\mathbf{g}$ is the vector of random genetic background effects, $\mathbf{W}$ is the SNP marker matrix, and $\mathbf{m}$ is the vector of fixed additive SNP effects. $\mathbf{g}$ is assumed to be distributed such that $\mathbf{g} \sim N(0, \mathbf{K} \sigma^2_g )$, where $\mathbf{K}$ models the population structure among genotypes.

The G and K models were fitted to scan for assocations.

Load the data, adjust p-values, and plot

```{r stability.gwas}

## Quality control
# Inspect the qvalue results

## QQ plot
qq_gwas_pheno_fw <- gwas_pheno_fw_adj %>% 
  arrange(., trait, model, p_value) %>% 
  mutate(p_exp = punif(q = seq(0, 1, length.out = n())), 
         neg_log10_p_exp = -log10(p_exp))

# Plot
g_qq <- qq_gwas_pheno_fw %>%
  ggplot(aes(x = neg_log10_p_exp, y = p_value_neg_log10, col = model)) + 
  geom_point() + 
  geom_abline(slope = 1, intercept = 0) + 
  facet_wrap(~ trait, ncol = 2) +
  labs(title = "QQ Plot", subtitle = "TP and VP") +
  theme_bw()

# save_file <- file.path(fig_dir, "gwas_pheno_fw_qq.jpg")
save_file <- file.path(fig_dir, "gwas_pheno_fw_qq_multi_geno.jpg")
ggsave(filename = save_file, plot = g_qq, width = 5, height = 7)



# Manhattan plot - use pagination by trait

g_man_tp <- gwas_pheno_fw_adj %>%
  mutate(chrom = as.factor(chrom)) %>% 
  # ggplot(aes(x = pos / 1000000, y = neg_log10_p_adj, group = chrom, col = chrom)) + 
  ggplot(aes(x = pos / 1000000, y = q_value_neg_log10, group = chrom, col = chrom)) +
  geom_point() + 
  geom_hline(aes(yintercept = neg_log10_fdr05, lty = "FDR 05%")) +
  geom_hline(aes(yintercept = neg_log10_fdr10, lty = "FDR 10%")) +
  facet_grid(trait + model ~ chrom, switch = "x", scales = "free", space = "free_x") +
  scale_color_manual(guide = FALSE, 
                     values = set_names(rep(c("black", "grey75"), length.out = 7), seq(1, 7))) +
  ylab("-log(q)") +
  xlab("Position (Mbp)") +
  labs(title = "Assocation Analysis", 
       caption = str_c("No. markers: ", ncol(M), "; No. lines: ", nrow(M))) +
  theme_bw() +
  theme(
    panel.spacing.x = unit(x = 0, units = "in"),
    panel.border = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )

# save_file <- file.path(fig_dir, "gwas_pheno_fw_man.jpg")
save_file <- file.path(fig_dir, "gwas_pheno_fw_man_multi_geno.jpg")
ggsave(filename = save_file, plot = g_man_tp, width = 7, height = 12)


## Only consider "G" models
g_man_tp_G <- gwas_pheno_fw_adj %>%
  filter(model %in% c("G", "QG")) %>%
  separate(trait, into = c("trait", "stability"), sep = "_") %>%
  mutate(chrom = as.factor(chrom),
         stability = if_else(stability == "b", "Linear", "Non-Linear")) %>% 
  # ggplot(aes(x = pos / 1000000, y = neg_log10_p_adj, group = chrom, col = chrom)) + 
  ggplot(aes(x = pos / 1000000, y = q_value_neg_log10, group = chrom, col = chrom)) +
  geom_point() + 
  geom_hline(aes(yintercept = neg_log10_fdr05, lty = "FDR 05%")) +
  geom_hline(aes(yintercept = neg_log10_fdr10, lty = "FDR 10%")) +
  facet_grid(trait + stability + model ~ chrom, switch = "x", scales = "free", space = "free_x") +
  scale_color_manual(guide = FALSE, 
                     values = set_names(rep(c("black", "grey75"), length.out = 7), seq(1, 7))) +
  scale_linetype_discrete(guide = FALSE) +
  ylab("-log(q)") +
  xlab("Position (Mbp)") +
  labs(title = "Genomewide Assocation Analysis for Phenotypic Stability",
       caption = str_c("No. markers: ", ncol(M), "; No. lines: ", nrow(M),
                       "\nLines indicate significance threshold at FDR = 5% and 10%")) +
  theme_bw() +
  theme(
    panel.spacing.x = unit(x = 0, units = "in"),
    panel.border = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )

# save_file <- file.path(fig_dir, "gwas_pheno_fw_manhattan_Gmodels.jpg")
save_file <- file.path(fig_dir, "gwas_pheno_fw_manhattan_Gmodels_multi_geno.jpg")
ggsave(filename = save_file, plot = g_man_tp_G, width = 10, height = 12)


## Only consider the G model
g_man_Gonly <- gwas_pheno_fw_adj %>%
  filter(model == "G") %>%
  separate(trait, into = c("trait", "stability"), sep = "_") %>%
  mutate(chrom = as.factor(chrom),
         stability = if_else(stability == "b", "Linear", "Non-Linear")) %>% 
  # filter(stability == "Type II") %>%
  ggplot(aes(x = pos / 1000000, y = q_value_neg_log10, group = chrom, col = chrom)) + 
  geom_point() + 
  geom_hline(aes(yintercept = neg_log10_fdr05, lty = "FDR 05%")) +
  geom_hline(aes(yintercept = neg_log10_fdr10, lty = "FDR 10%")) +
  facet_grid(trait + stability ~ chrom, switch = "x", scales = "free", space = "free_x") +
  scale_color_manual(guide = FALSE, 
                     values = set_names(rep(c("black", "grey75"), length.out = 7), seq(1, 7))) +
  scale_linetype_discrete(guide = FALSE) +
  ylab("-log(q)") +
  xlab("Position (Mbp)") +
  labs(title = "Genomewide Assocation Analysis for Phenotypic Stability",
       caption = str_c("No. markers: ", ncol(M), "; No. lines: ", nrow(M),
                       "\nLines indicate significance threshold at FDR = 5% and 10%")) +
  theme_bw() +
  theme(
    panel.spacing.x = unit(x = 0, units = "in"),
    panel.border = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )

# save_file <- file.path(fig_dir, "gwas_pheno_fw_man_Gonly.jpg")
save_file <- file.path(fig_dir, "gwas_pheno_fw_man_Gonly_multi_geno.jpg")
ggsave(filename = save_file, plot = g_man_Gonly, width = 9, height = 9, dpi = 1000)

```


Combine the mean and stability manhattan plots into a single graph


```{r plot.mean.and.stability}

# Combine mean and stability into a single df.
# Convert the stability -log10(q) values into their opposites (i.e. *-1)
gwas_geno_mean_use <- gwas_geno_mean_adj %>%
  ungroup() %>%
  filter(model == "G")  %>%
  select(marker:trait, contains("neg_log10_fdr"), Mean = q_value_neg_log10)

gwas_pheno_fw_use <- gwas_pheno_fw_adj %>%
  ungroup() %>%
  filter(model == "G") %>%
  separate(trait, into = c("trait", "stability"), sep = "_") %>%
  mutate(stability = if_else(stability == "b", "linear", "non-linear")) %>% 
  select(marker:trait, contains("neg_log10_fdr"), stability, Stability = q_value_neg_log10)

# Extract the palette for use in the manhattan plot
library(neyhart)
palette_use <- umn_palette("Secondary_Tier2")[c(3,8,4,9)]
odd_chrom <- seq(1, 7, by = 2)

gwas_use <-  gwas_pheno_fw_use %>%
  filter(stability == "linear") %>%
  full_join(gwas_geno_mean_use, .) %>% 
  gather(test, q_value_neg_log10, -marker:-neg_log10_fdr10, -stability) %>% 
  mutate_at(vars(q_value_neg_log10, contains("neg_log10_fdr")), funs(if_else(test == "Stability", . * -1, .))) %>%
  mutate(color = case_when(chrom %in% odd_chrom & test == "Mean" ~ "palette1",
                           !chrom %in% odd_chrom & test == "Mean" ~ "palette2",
                           chrom %in% odd_chrom & test == "Stability" ~ "palette3",
                           !chrom %in% odd_chrom & test == "Stability" ~ "palette4"),
         chrom = as.factor(chrom))

# color vector for ggplot
color_value <- set_names(palette_use, unique(gwas_use$color))

## Plot
g_mean_fw_man <- gwas_use %>%
  ggplot(aes(x = pos / 1000000, y = q_value_neg_log10)) + 
  geom_hline(yintercept = 0)+
  geom_point(aes(col = color)) +
  geom_hline(aes(yintercept = neg_log10_fdr05, lty = "FDR 05%")) +
  geom_hline(aes(yintercept = neg_log10_fdr10, lty = "FDR 10%")) +
  facet_grid(trait ~ chrom, switch = "x", scales = "free", space = "free_x") +
  scale_linetype_discrete(guide = FALSE) +
  scale_color_manual(values = color_value, labels = c("Mean", "Mean", "Stability", "Stability"), 
                     guide = FALSE) +
  ylab("-log(q)") +
  xlab("Position (Mbp)") +
  labs(title = "Genomewide Assocation Analyses") +
  theme_bw() +
  theme(panel.spacing.x = unit(x = 0, units = "cm"),
        panel.spacing.y = unit(x = 1, units = "cm"),
        # panel.border = element_rect(color = "grey75"),
        panel.border = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1),
        axis.text.y = element_blank())
  
save_file <- file.path(fig_dir, "gwas_pheno_mean_and_linear_stability.jpg")
ggsave(filename = save_file, plot = g_mean_fw_man, width = 9, height = 8)


## Plot non-linear stability

gwas_use <-  gwas_pheno_fw_use %>%
  filter(stability == "non-linear") %>%
  full_join(gwas_geno_mean_use, .) %>% 
  gather(test, q_value_neg_log10, -marker:-neg_log10_fdr10, -stability) %>% 
  mutate_at(vars(q_value_neg_log10, contains("neg_log10_fdr")), funs(if_else(test == "Stability", . * -1, .))) %>%
  mutate(color = case_when(chrom %in% odd_chrom & test == "Mean" ~ "palette1",
                           !chrom %in% odd_chrom & test == "Mean" ~ "palette2",
                           chrom %in% odd_chrom & test == "Stability" ~ "palette3",
                           !chrom %in% odd_chrom & test == "Stability" ~ "palette4"),
         chrom = as.factor(chrom))

## Plot
g_mean_fw_man <- gwas_use %>%
  ggplot(aes(x = pos / 1000000, y = q_value_neg_log10)) + 
  geom_hline(yintercept = 0)+
  geom_point(aes(col = color)) +
  geom_hline(aes(yintercept = neg_log10_fdr05, lty = "FDR 05%")) +
  geom_hline(aes(yintercept = neg_log10_fdr10, lty = "FDR 10%")) +
  facet_grid(trait ~ chrom, switch = "x", scales = "free", space = "free_x") +
  scale_linetype_discrete(guide = FALSE) +
  scale_color_manual(values = color_value, labels = c("Mean", "Mean", "Stability", "Stability"), 
                     guide = FALSE) +
  ylab("-log(q)") +
  xlab("Position (Mbp)") +
  labs(title = "Genomewide Assocation Analyses") +
  theme_bw() +
  theme(panel.spacing.x = unit(x = 0, units = "cm"),
        panel.spacing.y = unit(x = 1, units = "cm"),
        # panel.border = element_rect(color = "grey75"),
        panel.border = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1),
        axis.text.y = element_blank())

save_file <- file.path(fig_dir, "gwas_pheno_mean_and_nonlinear_stability.jpg")
ggsave(filename = save_file, plot = g_mean_fw_man, width = 9, height = 8)


```






#### Intervals around significant associations

First apply some filters to extract the significant SNPs 

```{r sig.overlap}

# Filter the main effect QTL and stability QTL for significant/suggestive markers
gwas_mean_interest_assoc <- gwas_geno_mean_adj %>%
  ungroup() %>% 
  filter(model == "G") %>% 
  mutate(interest = case_when(q_value_neg_log10 >= neg_log10_fdr05 ~ "significant", 
                              q_value_neg_log10 >= neg_log10_fdr10 ~ "suggestive", 
                              TRUE ~ "no_association"),
         test_type = "mean") %>%
  filter(interest != "no_association") %>%
  select(marker:trait, test_type, estimate, q_value, local_fdr, interest) %>%
  unnest(estimate)

gwas_fw_interest_assoc <- gwas_pheno_fw_adj %>%
  ungroup() %>% 
  separate(trait, c("trait", "stability_term"), sep = "_") %>%
  filter(model == "G") %>% 
  mutate(interest = case_when(q_value_neg_log10 >= neg_log10_fdr05 ~ "significant", 
                              q_value_neg_log10 >= neg_log10_fdr10 ~ "suggestive", 
                              TRUE ~ "no_association"),
         test_type = if_else(stability_term == "b", "linear_stability", "nonlinear_stability")) %>%
  filter(interest != "no_association") %>%
  select(marker:trait, test_type, estimate, q_value, local_fdr, interest) %>%
  unnest(estimate)

# Identify markers of interest for marker effect stability
mar_fw_linear_gtlt_interest <- mar_fw_linear_adj_gtlt %>% 
  ungroup() %>% 
  mutate(interest = case_when(neg_log_q >= neg_log10_fdr05 ~ "significant",
                              neg_log_q >= neg_log10_fdr10 ~ "suggestive",
                              TRUE ~ "no_association"),
         q_value = 10^-(neg_log_q),
         test_type = str_c(test_type, "_marker")) %>%
  filter(interest != "no_association") %>%
  select(marker, chrom, pos, trait, test_type, q_value, local_fdr, interest)

# Merge all dfs
sig_genomic_regions <- bind_rows(
  gwas_mean_interest_assoc,
  gwas_fw_interest_assoc,
  mar_fw_linear_gtlt_interest
)


# save
save_file <- file.path(result_dir, "S2MET_genomewide_test_sig_loci.RData")
save("sig_genomic_regions", file = save_file)


```




Now build intervals from each significant SNP for future overlap testing

```{r find.overlaps}

## How many significant or suggestive hits did we get per trait, per test?
n_sig_mar <- sig_genomic_regions %>% 
  group_by(trait, test_type, interest) %>% 
  summarize(n = n()) %>% 
  spread(interest, n) %>% 
  mutate(suggestive = ifelse(is.na(significant), suggestive, significant + suggestive))

# Plot
g_sig_mar <- n_sig_mar %>% 
  gather(interest, n, -trait:-test_type) %>% 
  ggplot(aes(x = test_type, y = n, fill = interest)) + 
  geom_col() + 
  facet_grid(trait ~. , scales = "free_y") +
  theme_bw() +
  theme(legend.position = "bottom",
        axis.title.y = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1))

save_file <- file.path(fig_dir, "number_significant_markers.jpg")
ggsave(filename = save_file, plot = g_sig_mar, height = 6, width = 4)

```


Establish criteria for clustering signficant SNPs as marking the same region.

Use an LD and distance threshold to create a distance matrix, then hierarchicaly cluster

```{r}

# LD threshold for grouping markers
LD_threshold <- 0.90
# Distance threshold for the clustering algorithm
bp_distance <- 1000000

# Window size for genomic region
wind_size <- 500000


# For each significant locus, find its LD with the next closest significant locus
sig_genomic_groups <- sig_genomic_regions %>% 
  filter(interest == "significant") %>%
  group_by(trait, test_type, chrom) %>%
  do({
    # Grab the data
    df <- .

    # If n_marker == 1, skip
    if (nrow(df) == 1) {
      new_grps1 <- 1

    } else {

      # Subset the genotype matrix and calculate LD
      LD <- cor(M[,df$marker])^2
      
      # Use a moving window approach to cluster SNPs based on LD
      ## Start with the first group, then try to add the next group
      ## If markers in the next group are not added, keep the groups
      ## If markers are added, use this larger group to try to grab the next markers
      
      # Create a distance matrix that combines the distance between markers and weighs them by the pairwise LD
      LD_dist <- as.dist(1 - LD)
      bp_dist <- dist(df$pos)
      
      clust_dist <- LD_dist * bp_dist
      
      # Assign new groups based on the clustering
      new_grps <- hclust(clust_dist) %>% 
        cutree(h = (1 - LD_threshold) * bp_distance)
      
      # If a group is repeated, cluster all SNPs in between those groups
      new_grps1 <- new_grps
      for (grp in unique(new_grps)) {
        
        # If the group is in the new vector, proceed
        if (grp %in% new_grps1) {
          which_grp <- which(new_grps == grp)
          new_grps1[seq(from = min(which_grp), to = max(which_grp))] <- grp
          
        }
      }
      
    }
      
    # Add the groups tp the df
    df %>% 
      mutate(new_groups = new_grps1)
    
  })

## Assign regions, while storing the marker names and adding windows around the
## range of positions
sig_genomic_regions <- sig_genomic_groups %>% 
  group_by(new_groups, add = T) %>% 
  mutate_at(vars(pos), funs(min, max)) %>% 
  group_by(test_type, min, max, add = T) %>%
  nest(marker, .key = "marker") %>%
  mutate(start = min - bp_distance,
         end = max + bp_distance) %>%
  select(trait:new_groups, start, end, marker)

# How many regions per chromosome, per trait, per test
n_regions <- sig_genomic_regions %>% 
  group_by(trait, test_type, chrom) %>% 
  summarize(n_interval = n_distinct(new_groups)) %>% 
  mutate(total_interval = sum(n_interval))

# Plot the number of intervals
n_regions %>% 
  ungroup() %>% 
  complete(trait, test_type) %>% 
  ggplot(aes(x = trait, y = total_interval, group = test_type, fill = test_type)) + 
  geom_col(position = "dodge") + 
  scale_fill_discrete(drop = FALSE)

g_n_regions <- n_regions %>% 
  ungroup() %>% 
  complete(trait, test_type) %>% 
  # filter(test_typea %in% c("main_effect_mean", "linear_stability", "non-linear_stability")) %>%
  ggplot(aes(x = trait, y = total_interval, group = test_type, fill = test_type)) + 
  geom_col(position = "dodge") + 
  scale_fill_discrete(drop = FALSE) +
  theme_bw()

save_file <- file.path(fig_dir, "number_significant_intervals.jpg")
ggsave(filename = save_file, plot = g_n_regions)

# sug_loci_df_intervals <- sug_loci_df_groups %>% 
#   group_by(group, add = T) %>% 
#   mutate(interval_start = min(start), interval_end = max(end)) %>%
#   ungroup() %>%
#   # Assign a y axis value to each test
#   mutate(test = factor(test, levels = c("mean", "stability", "stable", "sensitive")),
#          y_level = as.numeric(test))

## Plot the windows
g_sig_windows <- ggplot(data = S2TP_imputed_multi_genos_hmp, aes(x = pos / 1000000)) + 
  geom_segment(aes(x = min(pos  / 1000000), xend = max(pos  / 1000000), y = 0, yend = 0), lwd = 1.5) +
  geom_segment(data = mutate(sig_genomic_regions, y_level = as.numeric(as.factor(test_type))), 
               aes(x = start, xend = end, y = y_level, yend = y_level, col = test_type),
               lwd = 2.5, lineend = "round") +
  ylab("") +
  xlab("Position (Mbp)") +
  ylim(c(0, 10)) +
  facet_grid(trait ~ chrom, switch = "x", scales = "free_x", space = "free_x") +
  theme_bw() +
  theme(axis.text.y = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1),
        panel.spacing.x = unit(0, "in"),
        panel.border = element_rect(color = "grey75"))

save_file <- file.path(fig_dir, "significant_loci_windows.jpg")
ggsave(filename = save_file, plot = g_sig_windows, height = 7, width = 9)


# Zoom in chromosome 5 for heading date
g_sig_windows_HD <- S2TP_imputed_multi_genos_hmp %>%
  filter(chrom == 5, between(pos, 550000000, 650000000)) %>%
  ggplot(data = .) + 
  geom_segment(aes(x = min(pos  / 1000000), xend = max(pos  / 1000000), y = 0, yend = 0), lwd = 1.5) +
  geom_segment(data = mutate(sig_genomic_regions, y_level = as.numeric(as.factor(test_type))) %>%
                 filter(trait == "HeadingDate", chrom == 5, between(start, 550000000, 650000000)), 
               aes(x = start, xend = end, y = y_level, yend = y_level, col = test_type),
               lwd = 2.5, lineend = "round") +
  ylab("") +
  xlab("Position (Mbp)") +
  ylim(c(0, 6)) +
  xlim(c(550000000, 650000000)) +
  facet_grid(trait ~ chrom, switch = "x", scales = "free_x", space = "free_x") +
  theme_bw() +
  theme(axis.text.y = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1),
        panel.spacing.x = unit(0, "in"),
        panel.border = element_rect(color = "grey75"))

save_file <- file.path(fig_dir, "significant_loci_windows_zoom_HeadingDate.jpg")
ggsave(filename = save_file, plot = g_sig_windows_HD, height = 4, width = 6)



# ## Do the same for suggestive regions
# sug_loci_df_groups <- sig_loci_df %>% 
#   group_by(trait, test, chrom) %>%
#   do({
#     # Grab the data
#     df <- .
#     
#     # Add +/- 250 kbp to each position and assign preliminary groups based on overlap
#     df1 <- df %>% 
#       mutate(start = pos - 250000, end = pos + 250000,
#              prelim_group = IRanges(start = start, end = end) %>% 
#                findOverlaps(reduce(.), .) %>% queryHits() ) %>%
#       as.data.frame()
#     
#     # If n_marker == 1, skip
#     if (nrow(df1) == 1) {
#       new_grps <- 1
#       
#     } else {
#     
#       # Subset the genotype matrix and calculate LD
#       LD <- cor(M[,df1$marker])^2
#       # Extract the 'adjacent' LD
#       adj_LD <- diag(LD[,-1])
#       
#       # Vector of new groups
#       new_grps <- numeric(nrow(df1))
#       new_grps[df1$prelim_group == 1] <- 1
#       
#       # Iterate over preliminary groups in the df
#       for (grp in seq(2, max(df1$prelim_group))) {
#         # Extract the markers in that group plus the markers in previous group
#         grp_mar <- subset(df1, prelim_group == grp, marker, drop = TRUE)
#         prev_mar <- subset(df1, prelim_group == grp - 1, marker, drop = TRUE)
#         
#         # Are any markers in the previous group in high LD?
#         any_LD <- colSums(LD[grp_mar, prev_mar, drop = FALSE] >= LD_threshold)
#         # If any are, merge the groups
#         if (any(any_LD > 0)) {
#           # Change the grp number of those groups matching the previous group
#           # to the new group number
#           new_grps[new_grps == grp - 1] <- grp
#           
#         }
#         # Add the current grp
#         new_grps[df1$prelim_group == grp] <- grp
#       }
#       
#     }
# 
#       # Add groups to the markers
#       df1 %>% 
#         mutate(group = str_c("group", new_grps)) })





# # Plot for suggestive windows
# g_sug_windows <- ggplot(data = snp_info, aes(x = pos)) + 
#   geom_segment(aes(x = min(pos), xend = max(pos), y = 0, yend = 0), lwd = 1.5) +
#   geom_segment(data = sug_loci_df_intervals, 
#                aes(x = interval_start, xend = interval_end, y = y_level, yend = y_level, col = test),
#                lwd = 2.5, lineend = "round") +
#   ylim(c(0, 10)) +
#   facet_grid(trait ~ chrom, switch = "x", scales = "free_x", space = "free_x") +
#   theme_bw() +
#   theme(axis.text.y = element_blank(),
#         panel.spacing.x = unit(0, "in"),
#         panel.border = element_rect(color = "grey75"))
# 
# save_file <- file.path(fig_dir, "significant_loci_windows.jpg")
# ggsave(filename = save_file, plot = g_windows, height = 7, width = 9)


```








Window overlap

First find the overlap between mean and stability and quantify

Then find the overlap between stability and marker effect stability

```{r window.overlap}

# Convert the mean and stability intervals into GRanges
sig_intervals <- sig_loci_df_intervals %>% 
  droplevels() %>% 
  mutate(seqnames = str_c("chr", .$chrom, "H")) %>%
  distinct(trait, assoc_test, interval_start, interval_end, seqnames) %>%
  group_by(assoc_test, trait) %>% 
  do(range = GRanges(seqnames = .$seqnames, ranges = IRanges(start = .$interval_start, end = .$interval_end))) %>%
  select(trait, names(.)) %>%
  spread(assoc_test, range)



# Write a function that takes two lists of Ranges and finds the overlaps
# and number of overlaps
find_overlaps <- function(q, s) {
  # If null, return NA
  if (is.null(q) | is.null(s)) {
    data_frame(hits = list(NA), query_hits = list(NA), subject_hits = list(NA),
               query_length = as.integer(NA), subject_length = as.integer(NA),
               query_hits_n = as.integer(NA), subject_hits_n = as.integer(NA))
               
    
  } else {
    # Find overlaps
    hits <- suppressWarnings(findOverlaps(query = q, subject = s))
    # Data frame of query and subject hits, lengths, number
    data_frame(hits = list(hits)) %>% 
      mutate(query_hits = map(hits, ~q[queryHits(.),]), 
             subject_hits = map(hits, ~s[subjectHits(.),]),
             query_length = length(q), subject_length = length(s)) %>% 
      mutate_at(vars(contains("_hits")), funs(n = map_dbl(., length)))
  }
}



# For each trait, find the overlaps between mean and stability
interval_overlaps <- sig_intervals %>% 
  mutate(mean_linear_stability_overlap = list(main_effect_mean, linear_stability) %>% pmap(find_overlaps),
         mean_nonlinear_stability_overlap = list(main_effect_mean, `non-linear_stability`) %>% pmap(find_overlaps),
         linear_stability_mar_linear_stability_overlap = list(linear_stability, linear_marker_stability) %>% pmap(find_overlaps),
         linear_stability_mar_linear_plasticity_overlap = list(linear_stability, linear_marker_plasticity) %>% pmap(find_overlaps),
         nonlinear_stability_mar_nonlinear_stability_overlap = list(`non-linear_stability`, `non-linear_marker_plasticity`) %>% pmap(find_overlaps) )
    
# Extract the data.frame results
interval_overlaps_df <- interval_overlaps %>% 
  select(trait, contains("overlap")) %>% 
  gather(overlap, data, -trait) %>% 
  unnest() %>% 
  # Calculate proportions
  mutate(prop_query_hits = query_hits_n / query_length,
         prop_subject_hits = subject_hits_n / subject_length)

interval_overlaps_df %>% select(trait, overlap, query_hits_n:prop_subject_hits)

```


Gene models in the intervals

For the main effect and stability intervals, find gene annotations

```{r gene.ann}

# Load a venn diagram package
library(Vennerable)

# Load the gene annotation data
load("C:/Users/Jeff/Google Drive/Barley Lab/Projects/Genomics/Annotation/barley_genomic_ranges.RData")

# OVerlap the intervals and merge
sig_intervals_annotation <- sig_intervals %>% 
  gather(test, range, -trait) %>%
  filter(!map_lgl(range, is.null)) %>%
  mutate(overlap = map(range, ~mergeByOverlaps(query = ., subject = barley_grange_list$genes)))

# Save the cross-overlaps and the overlaps with the gene models
save_file <- file.path(result_dir, "genomewide_analysis_interval_overlaps.RData")
save("interval_overlaps_df", "sig_intervals_annotation", file = save_file)

  
# For each trait and test, extact a list if unique gene IDs
sig_intervals_gene_ids <- sig_intervals_annotation %>% 
  mutate(gene_ids = map(overlap, ~unique(.$gene_id)))

# Create set lists for Venn diagrams
sig_intervals_venn <- sig_intervals_gene_ids %>% 
  group_by(trait) %>% 
  do(gene_venn = set_names(.$gene_ids, .$test) %>% Venn)


# Plot a venn diagram of the gene ID overlaps between mean, stability, and marker stability
sig_intervals_venn$gene_venn %>% map(~plot(., doWeights = FALSE, type = if_else(NumberOfSets(.) < 4, "circles", "ellipses")))


```






## Appendix

### Archive

```{r}




###
# ### Archived code for testing the TP and TP + VP
# 
# # Tranpose the list
# gwas_pheno_fw1 <- transpose(gwas_pheno_fw)
# 
# # Add the model name to the df
# gwas_pheno_fw2 <- gwas_pheno_fw1 %>% 
#   map(~map_df(., ~mutate(.$scores, model = .$metadata$model) %>%
#                 filter(term == "main_effect")))
# 
# ## Adjust the p-values
# gwas_pheno_fw_adj <- gwas_pheno_fw2 %>%
#   map(~group_by(., trait, model) %>% 
#         mutate(p_adj = p.adjust(p_value, method = "fdr"),
#                q_value = qvalue(p = p_value)$qvalue,
#                neg_log10_fdr05 = -log10(0.05),
#                neg_log10_fdr10 = -log10(0.10)) %>%
#         mutate_at(vars(p_value, p_adj, q_value), funs(neg_log10 = -log10(.))) )
# 
# 
# ## Quality control
# ## QQ plot
# qq_gwas_pheno_fw_adj <- gwas_pheno_fw_adj %>% 
#   map(~arrange(., trait, model, p_value) %>% 
#         mutate(p_exp = punif(q = seq(0, 1, length.out = n())), 
#                neg_log10_p_exp = -log10(p_exp)))
# 
# # Plot
# g_qq_all <- qq_gwas_pheno_fw_adj$pop_all %>%
#   ggplot(aes(x = neg_log10_p_exp, y = p_value_neg_log10, col = model)) + 
#   geom_point() + 
#   geom_abline(slope = 1, intercept = 0) + 
#   facet_wrap(~ trait, ncol = 2) +
#   labs(title = "QQ Plot", subtitle = "TP and VP") +
#   theme_bw()
# 
# save_file <- file.path(fig_dir, "gwas_pheno_fw_qq_all.jpg")
# ggsave(filename = save_file, plot = g_qq_all, width = 5, height = 7)
# 
# 
# g_qq_tp <- qq_gwas_pheno_fw_adj$pop_tp %>%
#   ggplot(aes(x = neg_log10_p_exp, y = p_value_neg_log10, col = model)) + 
#   geom_point() + 
#   geom_abline(slope = 1, intercept = 0) + 
#   facet_wrap(~ trait, ncol = 2) +
#   labs(title = "QQ Plot", subtitle = "TP") +
#   theme_bw()
# 
# save_file <- file.path(fig_dir, "gwas_pheno_fw_qq_tp.jpg")
# ggsave(filename = save_file, plot = g_qq_tp, width = 5, height = 7)
# 
# 
# ## Manhattan plot
# g_man_all <- gwas_pheno_fw_adj$pop_all %>%
#   mutate(chrom = as.factor(chrom)) %>% 
#   # ggplot(aes(x = pos / 1000000, y = neg_log10_p_adj, group = chrom, col = chrom)) + 
#   ggplot(aes(x = pos / 1000000, y = q_value_neg_log10, group = chrom, col = chrom)) + 
#   geom_point() + 
#   geom_hline(aes(yintercept = neg_log10_fdr05, lty = "FDR 05%")) +
#   geom_hline(aes(yintercept = neg_log10_fdr10, lty = "FDR 10%")) +
#   facet_grid(trait + model ~ chrom, switch = "x", scales = "free", space = "free_x") +
#   scale_color_manual(values = set_names(rep(c("black", "grey75"), length.out = 7), seq(1, 7))) +
#   ylab("-log(q)") +
#   xlab("Position") +
#   labs(title = "Assocation Analysis", subtitle = "TP + VP") +
#   theme_bw() +
#   theme(
#     panel.spacing.x = unit(x = 0, units = "in"),
#     panel.border = element_blank()
#   )
# 
# save_file <- file.path(fig_dir, "gwas_pheno_fw_man_all.jpg")
# ggsave(filename = save_file, plot = g_man_all, width = 7, height = 12)
# 
# 
# g_man_tp <- gwas_pheno_fw_adj$pop_tp %>%
#   mutate(chrom = as.factor(chrom)) %>% 
#   # ggplot(aes(x = pos / 1000000, y = neg_log10_p_adj, group = chrom, col = chrom)) + 
#   ggplot(aes(x = pos / 1000000, y = q_value_neg_log10, group = chrom, col = chrom)) +
#   geom_point() + 
#   geom_hline(aes(yintercept = neg_log10_fdr05, lty = "FDR 05%")) +
#   geom_hline(aes(yintercept = neg_log10_fdr10, lty = "FDR 10%")) +
#   facet_grid(trait + model ~ chrom, switch = "x", scales = "free", space = "free_x") +
#   scale_color_manual(guide = FALSE, 
#                      values = set_names(rep(c("black", "grey75"), length.out = 7), seq(1, 7))) +
#   ylab("-log(q)") +
#   xlab("Position (Mbp)") +
#   labs(title = "Assocation Analysis", subtitle = "TP") +
#   theme_bw() +
#   theme(
#     panel.spacing.x = unit(x = 0, units = "in"),
#     panel.border = element_blank(),
#     axis.text.x = element_text(angle = 45, hjust = 1)
#   )
# 
# save_file <- file.path(fig_dir, "gwas_pheno_fw_man_tp.jpg")
# ggsave(filename = save_file, plot = g_man_tp, width = 7, height = 12)
# 
# 
# ## Only use the TP and only consider the G model
# g_man_tp <- gwas_pheno_fw_adj$pop_tp %>%
#   filter(model == "G") %>%
#   separate(trait, into = c("trait", "stability"), sep = "_") %>%
#   mutate(chrom = as.factor(chrom),
#          stability = if_else(stability == "b", "Type II", "Type III")) %>% 
#   # ggplot(aes(x = pos / 1000000, y = neg_log10_p_adj, group = chrom, col = chrom)) + 
#   ggplot(aes(x = pos / 1000000, y = q_value_neg_log10, group = chrom, col = chrom)) +
#   geom_point() + 
#   geom_hline(aes(yintercept = neg_log10_fdr05, lty = "FDR 05%")) +
#   geom_hline(aes(yintercept = neg_log10_fdr10, lty = "FDR 10%")) +
#   facet_grid(trait + stability ~ chrom, switch = "x", scales = "free", space = "free_x") +
#   scale_color_manual(guide = FALSE, 
#                      values = set_names(rep(c("black", "grey75"), length.out = 7), seq(1, 7))) +
#   scale_linetype_discrete(guide = FALSE) +
#   ylab("-log(q)") +
#   xlab("Position (Mbp)") +
#   labs(title = "Genomewide Assocation Analysis for Phenotypic Stability",
#        caption = "n = 183 lines used to calculate stability coefficients and n = 175 lines used\nfor association analysis. The bold line indicates the genomewide FDR threshold at 5%;\nthe dashed line at 10%.") +
#   theme_bw() +
#   theme(
#     panel.spacing.x = unit(x = 0, units = "in"),
#     panel.border = element_blank(),
#     axis.text.x = element_text(angle = 45, hjust = 1)
#   )
# 
# save_file <- file.path(fig_dir, "gwas_pheno_fw_man_tp_final.jpg")
# ggsave(filename = save_file, plot = g_man_tp, width = 10, height = 8)
# 
# 
# 
# ## Only use the TP and only consider the G model, plus only look at the slope from FW
# g_man_tp_typeii <- gwas_pheno_fw_adj$pop_tp %>%
#   filter(model == "G") %>%
#   separate(trait, into = c("trait", "stability"), sep = "_") %>%
#   mutate(chrom = as.factor(chrom),
#          stability = if_else(stability == "b", "Type II", "Type III")) %>% 
#   filter(stability == "Type II") %>%
#   ggplot(aes(x = pos / 1000000, y = q_value_neg_log10, group = chrom, col = chrom)) + 
#   geom_point() + 
#   geom_hline(aes(yintercept = neg_log10_fdr05, lty = "FDR 05%")) +
#   geom_hline(aes(yintercept = neg_log10_fdr10, lty = "FDR 10%")) +
#   facet_grid(trait ~ chrom, switch = "x", scales = "free", space = "free_x") +
#   scale_color_manual(guide = FALSE, 
#                      values = set_names(rep(c("black", "grey75"), length.out = 7), seq(1, 7))) +
#   scale_linetype_discrete(guide = FALSE) +
#   ylab("-log(q)") +
#   xlab("Position (Mbp)") +
#   labs(title = "Genomewide Assocation Analysis for Phenotypic Stability",
#        caption = "n = 183 lines used to calculate stability coefficients and n = 175 lines used\nfor association analysis. The bold line indicates the genomewide FDR threshold at 5%;\nthe dashed line at 10%.") +
#   theme_bw() +
#   theme(
#     panel.spacing.x = unit(x = 0, units = "in"),
#     panel.border = element_blank(),
#     axis.text.x = element_text(angle = 45, hjust = 1)
#   )
# 
# save_file <- file.path(fig_dir, "gwas_pheno_fw_man_tp_typeii.jpg")
# ggsave(filename = save_file, plot = g_man_tp_typeii, width = 9, height = 7, dpi = 1000)




```



