---
title: "S2 MET GWAS Analysis"
output: html_notebook
bibliography: C:/Users/Jeff/Documents/Literature/library.bib
---

This notebook will outline the analysis of GWAS mapping procedures using the S2MET data.

The GWAS will examine main effect QTL, QTLxE, and QTL influencing stability.


## Introduction

Load packages and set directories

```{r setup}

library(tidyverse)
library(readxl)
library(lme4)
library(rrBLUP)
library(broom)

# Project and other directories
source("C:/Users/Jeff/Google Drive/Barley Lab/Projects/S2MET_Mapping/source.R")

load(file.path(geno_dir, "S2_genos_hmp.RData"))

# Filter environments for those in which the TP was observed
S2_MET_BLUEs_use <- S2_MET_BLUEs %>% 
  group_by(trait, environment) %>%
  filter(sum(line_name %in% tp) > 1) %>%
  ungroup()

```


## Results

### Population Structure

First examine population structure within the germplasm

```{r pop.str}

# Calculate the A matrix
A_all <- A.mat(X = s2_imputed_mat_use, min.MAF = 0, max.missing = 1)
A_tp <- A.mat(X = s2_imputed_mat_use[tp_geno,], min.MAF = 0, max.missing = 1)

A_list <- list(all = A_all, tp = A_tp)

## Visualize pop structure
A_prcomp <- map(A_list, prcomp)

# Extract the proportion explained by each PC
lambda <- map(A_prcomp, "sdev") %>% map(~round(. / sum(.), 3) * 100)

# Convert to data.frame
A_prcomp_df_list <- map(A_prcomp, "rotation") %>% 
  map(~as.data.frame(.) %>% rownames_to_column("line_name"))

A_prcomp_df <- A_prcomp_df_list %>%
  map(~left_join(., entry_list, by = c("line_name" = "Line")) %>%
        select(line_name, program = Program, parent = `Parent?`, starts_with("PC")) %>%
        mutate(program = if_else(str_detect(line_name, "^2MS"), "M2", program))) %>%
  list(., c("All", "TP")) %>% 
  pmap(~mutate(.x, population = .y)) %>% 
  bind_rows()


# Plot
(g_pop_str <- A_prcomp_df %>% 
  ggplot(aes(x = PC1, y = PC2, col = program)) + 
  geom_point(size = 2) + 
  ylab("PC2") +
  xlab("PC1") +
  scale_color_discrete(guide = guide_legend(title = "Breeding\nProgram")) +
  facet_grid(~ population) +
  labs(
    title = "Population Structure of the S2MET Genotypes",
    subtitle = "The 'M2' program includes genotypes derived\nfrom crosses made among genotypes from the other programs."
  ) +
  theme_bw())

# Save this
save_file <- file.path(fig_dir, "population_structure.jpg")
ggsave(filename = save_file, plot = g_pop_str, width = 6, height = 5)

```


How is population structure correlated with the traits?

First, fit a model to calculate BLUEs for each genotype across all environments

```{r mme.blues}

trait_BLUEs <- S2_MET_BLUEs_use %>%
  group_by(trait) %>%
  do({
    df <- .
    # Standard errors as weights
    wts <- df$std_error^2
    
    # Control the model
    lmer_control <- lmerControl(check.nobs.vs.nlev = "ignore", check.nobs.vs.nRE = "ignore")
    
    # Set the contrasts of line_name and environment
    df_use <- df %>%
      mutate_at(vars(line_name, environment), as.factor)
    
    # Fit the model
    fit <- lmer(value ~ -1 + line_name + environment + (1|line_name:environment), data = df_use,
                weights = wts, control = lmer_control)
    
    # Extract BLUEs and return
    tidy(fit) %>% 
      filter(str_detect(term, "line_name"), group == "fixed") %>% 
      mutate(term = str_replace(term, "line_name", "")) %>% 
      select(line_name = term, value = estimate) })

# Combine the trait and PCs, duplicating the BLUEs to make summarizing smoother
trait_pop_str <- A_prcomp_df %>% 
  select(line_name:parent, population, PC1:PC3) %>% 
  gather(PC, eigenvalue, -line_name:-population) %>% 
  left_join(., trait_BLUEs, by = "line_name")

# Correlate trait BLUEs with eigenvectors
trait_pop_str_summ <- trait_pop_str %>% 
  group_by(population, PC, trait) %>% 
  summarize(trait_str_cor = cor(eigenvalue, value))


# Plot
(g_trait_pop_str <- trait_pop_str %>% 
  ggplot(aes(x = eigenvalue, y = value)) +
  geom_smooth(method = "lm", se = FALSE) +
  geom_point(aes(col = program)) +
  facet_grid(trait + population ~ PC, scales = "free") +
  ylab("Genotypic Value") +
  xlab("Eigenvalue") +
  labs(
    title = "Relationship Between Trait and Population Structure"
  ))

save_file <- file.path(fig_dir, "trait_pop_str_relationship.jpg")
ggsave(filename = save_file, plot = g_trait_pop_str, width = 6, height = 8, dpi = 500)

```




### GWAS

#### Stability QTL

First we will map QTL for trait stability. The stability coefficients are first calculated using FW regression, and the coefficients are then used in the GWAS mixed model:

$$
y_{i} = \mu + \sum^m_{m=1} x_{im} \alpha_m + g_{i} + \epsilon_{i}
$$

which models the stability of genotype *i* as the sum of the grand mean, the sum of the marker allele effects carried by genotype *i*, the residual genetic effect of genotype *i*, and the residual effect associated with the observation.

In vector notation this becomes:


$$
\mathbf{y = X \beta + Wm + Zu + e}
$$

where $\mathbf{y}$ is the response vector, $\mathbf{X}$ is the fixed effect incidence matrix modeling environment as a fixed effect, $\mathbf{\beta}$ is the vector of fixed environmental effects, $\mathbf{Z}$ is the incidence matrix modeling the random genetic background effect, $\mathbf{g}$ is the vector of random genetic background effects, $\mathbf{W}$ is the SNP marker matrix, and $\mathbf{m}$ is the vector of fixed additive SNP effects. $\mathbf{g}$ is assumed to be distributed such that $\mathbf{g} \sim N(0, \mathbf{K} \sigma^2_g )$, where $\mathbf{K}$ models the population structure among genotypes.

I used the `GWAS()` function in the `rrBLUP` package to fit this model. The first round used the P + K model, and the second round used the K only model.


#### Main Effect QTL and QTLxE

The above model does not account for QTLxE interaction. We will therefore split up the modeling of main effect QTL, QTLxE, and stability QTL.

We fit the following model for QTL main effects and QTLxE

$$
y_{ij} = \mu + E_j + \sum^m_{m=1} x_{im} (\alpha_m + \delta_{mj}) + g_{ij} + \epsilon_{ij}
$$

which models the mean of genotype *i* in environment *j* as the sum of the grand mean, the effect of the *j*th environment, the sum of the marker allele effects ($\alpha_m$) and environment-specific marker deviations ($\delta_{mj}$) carried by genotype *i*, the residual genetic effect of genotype *i* in environment *j*, and the residual effect associated with the observation.

In vector notation this becomes:


$$
\mathbf{y = X \beta + Wm + Vd + Zu + e}
$$

where $\mathbf{y}$ is the response vector, $\mathbf{X}$ is the fixed effect incidence matrix modeling environment as a fixed effect, $\mathbf{\beta}$ is the vector of fixed environmental effects, $\mathbf{Z}$ is the incidence matrix modeling the random genetic background effect, $\mathbf{g}$ is the vector of random genetic background effects, $\mathbf{W}$ is the SNP marker matrix, $\mathbf{m}$ is the vector of fixed additive SNP effects, $\mathbf{V}$ is the incidence matrix of SNPs across environments, and $\mathbf{d}$ is the vector of environment-specific SNP deviations. $\mathbf{g}$ is assumed to be distributed such that $\mathbf{g} \sim N(0, \mathbf{K} \sigma^2_g )$, where $\mathbf{K}$ models the population structure among genotypes.

The G and K models were fitted to scan for assocations.

Load the data, adjust p-values, and plot

```{r stability.gwas}

# Load the data
load(file.path(result_dir, "S2MET_pheno_fw_gwas_results.RData"))

# Tranpose the list
gwas_pheno_fw1 <- transpose(gwas_pheno_fw)

# Add the model name to the df
gwas_pheno_fw2 <- gwas_pheno_fw1 %>% 
  map(~map_df(., ~mutate(.$scores, model = .$metadata$model) %>%
                filter(term == "main_effect")))

## Adjust the p-values
gwas_pheno_fw_adj <- gwas_pheno_fw2 %>%
  map(~group_by(., trait, model) %>% 
        mutate(p_adj = p.adjust(p_value, method = "fdr"),
               neg_log10_p = -log10(p_value),
               neg_log10_p_adj = -log10(p_adj),
               neg_log10_fdr05 = -log10(0.05),
               neg_log10_fdr10 = -log10(0.10)))


## Quality control
## QQ plot
qq_gwas_pheno_fw_adj <- gwas_pheno_fw_adj %>% 
  map(~arrange(., trait, model, p_value) %>% 
        mutate(p_exp = punif(q = seq(0, 1, length.out = n())), 
               neg_log10_p_exp = -log10(p_exp)))

# Plot
g_qq_all <- qq_gwas_pheno_fw_adj$pop_all %>%
  ggplot(aes(x = neg_log10_p_exp, y = neg_log10_p, col = model)) + 
  geom_point() + 
  geom_abline(slope = 1, intercept = 0) + 
  facet_wrap(~ trait, ncol = 2) +
  labs(title = "QQ Plot", subtitle = "TP and VP") +
  theme_bw()

save_file <- file.path(fig_dir, "gwas_pheno_fw_qq_all.jpg")
ggsave(filename = save_file, plot = g_qq_all, width = 5, height = 7)


g_qq_tp <- qq_gwas_pheno_fw_adj$pop_tp %>%
  ggplot(aes(x = neg_log10_p_exp, y = neg_log10_p, col = model)) + 
  geom_point() + 
  geom_abline(slope = 1, intercept = 0) + 
  facet_wrap(~ trait, ncol = 2) +
  labs(title = "QQ Plot", subtitle = "TP") +
  theme_bw()

save_file <- file.path(fig_dir, "gwas_pheno_fw_qq_tp.jpg")
ggsave(filename = save_file, plot = g_qq_tp, width = 5, height = 7)


## Manhattan plot
g_man_all <- gwas_pheno_fw_adj$pop_all %>%
  mutate(chrom = as.factor(chrom)) %>% 
  ggplot(aes(x = pos / 1000000, y = neg_log10_p_adj, group = chrom, col = chrom)) + 
  geom_point() + 
  geom_hline(aes(yintercept = neg_log10_fdr05, lty = "FDR 05%")) +
  geom_hline(aes(yintercept = neg_log10_fdr10, lty = "FDR 10%")) +
  facet_grid(trait + model ~ chrom, switch = "x", scales = "free") +
  scale_color_manual(values = set_names(rep(c("black", "grey75"), length.out = 7), seq(1, 7))) +
  ylab("-log(p)") +
  xlab("Position") +
  labs(title = "Assocation Analysis", subtitle = "TP + VP") +
  theme_bw() +
  theme(
    panel.spacing.x = unit(x = 0, units = "in"),
    panel.border = element_blank()
  )

save_file <- file.path(fig_dir, "gwas_pheno_fw_man_all.jpg")
ggsave(filename = save_file, plot = g_man_all, width = 7, height = 12)


g_man_tp <- gwas_pheno_fw_adj$pop_tp %>%
  mutate(chrom = as.factor(chrom)) %>% 
  ggplot(aes(x = pos / 1000000, y = neg_log10_p_adj, group = chrom, col = chrom)) + 
  geom_point() + 
  geom_hline(aes(yintercept = neg_log10_fdr05, lty = "FDR 05%")) +
  geom_hline(aes(yintercept = neg_log10_fdr10, lty = "FDR 10%")) +
  facet_grid(trait + model ~ chrom, switch = "x", scales = "free") +
  scale_color_manual(guide = FALSE, 
                     values = set_names(rep(c("black", "grey75"), length.out = 7), seq(1, 7))) +
  ylab("-log(p)") +
  xlab("Position (Mbp)") +
  labs(title = "Assocation Analysis", subtitle = "TP") +
  theme_bw() +
  theme(
    panel.spacing.x = unit(x = 0, units = "in"),
    panel.border = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )

save_file <- file.path(fig_dir, "gwas_pheno_fw_man_tp.jpg")
ggsave(filename = save_file, plot = g_man_tp, width = 7, height = 12)



```







First, association for the main effect of QTL and QTLxE. We first look at the model comparisons for assocation analyses using the TP only

```{r qtlxe.model}

# Load the data
load(file.path(result_dir, "/S2MET_gwas_tp_model_comparison.RData"))

## Add the model name to the dfs
gwas_tp_main_df <- gwas_tp_main %>% 
  map_df(~mutate(.$scores, model = .$metadata$model)) %>% 
  as_data_frame()


# Correct p-values for multiple testing
gwas_tp_main_df_adj <- gwas_tp_main_df %>% 
  group_by(trait, term, model) %>% 
  mutate(p_adj = p.adjust(p = p_value, method = "fdr"),
         neg_log_p = -log10(p_value),
         neg_log_p_adj = -log10(p_adj),
         neg_log_fdr05 = -log10(0.05),
         neg_log_fdr10 = -log10(0.10)) %>%
  ungroup()

# Q-Q Plot per trait
gwas_tp_qq <- gwas_tp_main_df_adj %>% 
  group_by(trait, term, model) %>% 
  arrange(p_value) %>% 
  mutate(exp_p_value = punif(q = seq(0, 1, length.out = n())),
         neg_log_exp_p_value = -log10(exp_p_value))
  

(g_gwas_tp_qq <- gwas_tp_qq %>%
  ggplot(aes(x = neg_log_exp_p_value, y = neg_log_p, col = model)) +
  geom_point() +
  geom_abline(intercept = 0, slope = 1, col = "red", lwd = 1) +
  facet_grid(trait ~ term, scales = "free_y") +
  ylab("Observed -log10(p)") +
  xlab("Expected -log10(p)") +
  labs(
    title = "Q-Q Plot of Association Results"
  ) +
  theme_bw() )

# Save this
save_file <- file.path(fig_dir, "gwas_qq.jpg")
ggsave(filename = save_file, plot = g_gwas_qq, width = 6, height = 6)



```

Load data for the genome scan for main effect and QxE



Manhattan plot

```{r plot.gwas.main.qxe}

# Results for the TP only
load(file.path(result_dir, "S2MET_gwas_tp_main_qxe_model_comparison.RData"))

## Add the model name to the dfs
gwas_tp_main_qxe_df <- gwas_tp_main_qxe %>% 
  map_df(~mutate(.$scores, model = .$metadata$model)) %>% 
  as_data_frame()

# Correct p-values for multiple testing
gwas_tp_main_qxe_adj <- gwas_tp_main_qxe_df %>% 
  group_by(trait, term, model) %>% 
  mutate(p_adj = p.adjust(p = p_value, method = "fdr"),
         neg_log_p = -log10(p_value),
         neg_log_p_adj = -log10(p_adj),
         neg_log_fdr05 = -log10(0.05),
         neg_log_fdr10 = -log10(0.10)) %>%
  ungroup()

# Q-Q Plot per trait
gwas_tp_main_qxe_qq <- gwas_tp_main_qxe_adj %>% 
  group_by(trait, term, model) %>% 
  arrange(p_value) %>% 
  mutate(exp_p_value = punif(q = seq(0, 1, length.out = n())),
         neg_log_exp_p_value = -log10(exp_p_value))

(gwas_tp_main_qxe_qq_plot <- gwas_tp_main_qxe_qq %>%
  ggplot(aes(x = neg_log_exp_p_value, y = neg_log_p, col = model)) +
  geom_point() +
  geom_abline(intercept = 0, slope = 1, col = "red", lwd = 1) +
  facet_grid(trait ~ term, scales = "free_y") +
  ylab("Observed -log10(p)") +
  xlab("Expected -log10(p)") +
  labs(
    title = "Q-Q Plot of Association Results"
  ) +
  theme_bw() )

## Manhattan plot
(gwas_tp_main_qxe_man <- gwas_tp_main_qxe_adj %>%
  mutate(chrom = as.factor(chrom),
         pos = pos / 1000000) %>%
  ggplot(aes(x = pos, y = neg_log_p_adj, col = chrom)) +
  geom_point() +
  geom_hline(aes(yintercept = neg_log_fdr05, lty = "FDR 05%")) +
  geom_hline(aes(yintercept = neg_log_fdr10, lty = "FDR 10%")) +
  facet_grid(trait + term + model ~ chrom, space = "free_x", scale = "free", switch = "x") +
  ylab("Adjusted -log(p)") +
  xlab("Position (Mbp)") +
  scale_color_brewer(palette = "Set2", drop = FALSE, guide = FALSE) +
  scale_linetype(guide = guide_legend(title = NULL)) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    panel.border = element_blank(),
    panel.spacing.x = unit(0, units = "in")
  ) +
  labs(
    title = "GWAS of Main Effect QTL and QxE"
  ))



```











```{r}


# Plot the GWAS results
(g_qtlxe_gwas <- gwas_qtlxe_out2 %>%
  mutate(chrom = as.factor(chrom),
         pos = pos / 1000000) %>%
  ggplot(aes(x = pos, y = neg_log_p_adj, col = chrom)) +
  geom_point() +
  geom_hline(aes(yintercept = neg_log_fdr05, lty = "FDR 05%")) +
  geom_hline(aes(yintercept = neg_log_fdr10, lty = "FDR 10%")) +
  facet_grid(trait + term ~ chrom, space = "free_x", scale = "free", switch = "x") +
  ylab("Adjusted -log(p)") +
  xlab("Position (Mbp)") +
  scale_color_brewer(palette = "Set2", drop = FALSE, guide = FALSE) +
  scale_linetype(guide = guide_legend(title = NULL)) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    panel.border = element_blank()
  ) +
  labs(
    title = "GWAS of Genotypic Main Effects and Sensitivity Coefficients"
  ))


# Save this
save_file <- file.path(fig_dir, "gwas_fw_results.jpg")
ggsave(filename = save_file, plot = g_fw_gwas, width = 12, height = 7, dpi = 500)


```








```{r main.model1}

# Load the data
load(file.path(result_dir, "S2MET_gwas_fw_results.RData"))

# Gather data
gwas_main_out1 <- gwas_fw_out %>% 
  gather(trait, neg_log_p, -marker:-pos) %>%
  # Convert -log_p to p
  mutate(p_value = 10 ^ (-1 * neg_log_p))

# Correct the p_values for an FDR of 0.05
gwas_main_out2 <- gwas_main_out1 %>%
  group_by(trait) %>%
  mutate(p_adj = p.adjust(p = p_value, method = "fdr"),
         neg_log_p_adj = -log10(p_adj),
         neg_log_fdr05 = -log10(0.05),
         neg_log_fdr10 = -log10(0.10)) %>%
  ungroup() %>%
  separate(trait, c("trait", "coef"), sep = "_") %>%
  mutate(coef = str_replace_all(coef, c(g = "GenotypeMean", b = "LinearStability", delta = "NonLinearStability")))


# Q-Q Plot per trait
gwas_qq <- gwas_main_out2 %>%
  group_by(trait, coef) %>%
  arrange(trait, coef, p_value) %>% 
  mutate(exp_neg_log_p = seq_along(neg_log_p), 
         exp_neg_log_p = -(log10(exp_neg_log_p / (length(exp_neg_log_p)+1)))) %>%
  arrange(trait, coef, chrom, pos)

(g_gwas_qq <- gwas_qq %>%
  ggplot(aes(x = exp_neg_log_p, y = neg_log_p)) +
  geom_point() +
  geom_abline(intercept = 0, slope = 1, col = "red", lwd = 1) +
  facet_grid(trait ~ coef) +
  ylab("Observed -log10(p)") +
  xlab("Expected -log10(p)") +
  labs(
    title = "Q-Q Plot of Association Results"
  ) +
  theme_bw() )

# Save this
save_file <- file.path(fig_dir, "gwas_qq.jpg")
ggsave(filename = save_file, plot = g_gwas_qq, width = 6, height = 6)


```


```{r plot.gwas}

# Plot the GWAS results
(g_fw_gwas <- gwas_main_out2 %>%
  mutate(chrom = as.factor(chrom),
         pos = pos / 1000000) %>%
  ggplot(aes(x = pos, y = neg_log_p_adj, col = chrom)) +
  geom_point() +
  geom_hline(aes(yintercept = neg_log_fdr05, lty = "FDR 05%")) +
  geom_hline(aes(yintercept = neg_log_fdr10, lty = "FDR 10%")) +
  facet_grid(trait + coef ~ chrom, space = "free_x", scale = "free", switch = "x") +
  ylab("Adjusted -log(p)") +
  xlab("Position (Mbp)") +
  scale_color_brewer(palette = "Set2", drop = FALSE, guide = FALSE) +
  scale_linetype(guide = guide_legend(title = NULL)) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    panel.border = element_blank()
  ) +
  labs(
    title = "GWAS of Genotypic Main Effects and Sensitivity Coefficients"
  ))


# Save this
save_file <- file.path(fig_dir, "gwas_fw_results.jpg")
ggsave(filename = save_file, plot = g_fw_gwas, width = 12, height = 7, dpi = 500)


```


Results of GWAS for stability to environmental covariables

```{r ec.stability.gwas}

# Load the data
load(file.path(result_dir, "S2MET_gwas_fw_ec_one_year_results.RData"))




```










