---
title: "S2 MET Marker Stability Analysis"
output: html_notebook
bibliography: C:/Users/Jeff/Documents/Literature/library.bib
---

This notebook will outline the analysis of GWAS mapping procedures using the S2MET data.


## Introduction

Load packages and set directories

```{r setup}

library(tidyverse)
library(stringr)
library(readxl)
library(modelr)
library(rrBLUP)
library(GGally)
library(GenomicRanges)
library(qvalue)
library(purrrlyr)

# Project and other directories
source("C:/Users/Jeff/Google Drive/Barley Lab/Projects/S2MET_Mapping/source.R")

# Load the marker effect FW results
load(file.path(result_dir, "S2MET_pheno_mar_eff_fw_results.RData"))
# Load the stability results
load(file.path(result_dir, "S2MET_pheno_fw_regression_results.RData"))

tp_geno <- tp_geno_multi

# Filter the BLUEs to use
S2_MET_BLUEs_use <- S2_MET_BLUEs %>% 
  filter(line_name %in% c(tp_geno)) %>%
  droplevels()

M <- S2TP_imputed_multi_genos_mat[tp_geno,]

# Quartile coefficient of dispersion statistics
qcd <- function(x) {
  quants <- quantile(x = x, probs = c(0.50, 0.75))
  as.numeric(diff(quants) / sum(quants))
}

```




## Results

First group markers into those that are highly stable, highly sensitive, or neither

```{r grp.mars}

# Get the distinct observations of the stability measures
S2_MET_marker_eff_pheno_fw_uniq <- S2_MET_marker_eff_pheno_fw %>% 
  select(trait, marker:pos, b_std_error, df, stability_term, estimate) %>% 
  distinct()

# What should be the significance level
alpha <- 0.05

# For each trait, calculate empirical thresholds for significance
S2_MET_marker_eff_pheno_fw_sig <- S2_MET_marker_eff_pheno_fw_uniq %>%
  filter(stability_term == "b") %>% 
  group_by(trait) %>% 
  # mutate(estimate = scale(estimate)) %>%
  mutate(lower_perc = quantile(estimate, alpha / 2), 
         upper_perc = quantile(estimate, 1 - (alpha / 2))) %>%
  ungroup() %>%
  mutate(significance = case_when(estimate >= upper_perc ~ "sensitive",
                                  estimate <= lower_perc ~ "stable",
                                  TRUE ~ "not_significant"),
         marker_type = if_else(str_detect(marker, "^S"), "GBS", "BOPA"))

# Calculate a table of GBS markers and sensitive/stable/notsignificant
mar_stab_table <- S2_MET_marker_eff_pheno_fw_sig %>% 
  group_by(trait, marker_type, significance) %>% 
  summarize(n = n()) %>% 
  mutate(prop = n / sum(n))

mar_stab_table %>% 
  select(-n) %>% 
  spread(significance, prop)


# Plot the distributions of marker effect stabilities and show the empirical threholds
S2_MET_marker_eff_pheno_fw_sig %>% 
  ggplot(aes(x = estimate, fill = "red")) + 
  geom_density() + 
  geom_vline(aes(xintercept = upper_perc)) +
  geom_vline(aes(xintercept = lower_perc)) + 
  facet_grid(trait ~ .) + 
  scale_fill_discrete(guide = FALSE) + 
  theme_bw()

# Calculate quartile coefficients of dispersion
S2_MET_marker_eff_pheno_fw_uniq %>% 
  mutate(stability_term = if_else(stability_term == "b", "linear", "non-linear")) %>%
  group_by(trait, stability_term) %>% 
  summarize(qcd = qcd(estimate)) %>%
  ggplot(aes(x = trait, y = qcd, shape = stability_term)) +
  geom_point(size = 3) +
  theme_bw()

# Boxplots
S2_MET_marker_eff_pheno_fw_uniq %>% 
  mutate(stability_term = if_else(stability_term == "b", "linear", "non-linear")) %>% 
  ggplot(aes(x = trait, y = estimate)) + 
  geom_boxplot() +
  facet_wrap(~ stability_term, scales = "free_y")


```


Determine the proportion of GxE variance ($V_{GE}$) that is explained by sensitive/stable markers vs not signficant markers

The mixed-model looks like:

$$
y_{ij} = \mu + g_i + t_j + (g_st)_{ij} + (g_{ns}t)_{ij} + \epsilon_{ij},
$$

where $g_i \sim N(0, \mathbf{K}\sigma^2_g)$, $(g_st)_{ij} \sim N(0, \mathbf{K_s}\sigma^2_{g_st})$, and $(g_{ns}t)_{ij} \sim N(0, \mathbf{K_{ns}}\sigma^2_{g_{ns}t})$. $\mathbf{K}$ is the kinship matrix calculated using all markers, $\mathbf{K_s}$ is the kinship matrix calculated using only stable/sensitive markers, and $\mathbf{K_{ns}}$ is the kinship matrix calculated using only non-signficiant markers

```{r prop.gxe1}

# Load the results
load(file.path(result_dir, "S2MET_pheno_mar_fw_varcomp_GrainYield.RData"))

var_comp_prop <- var_comp_out$GrainYield %>% 
  bind_rows() %>%
  by_row(sum, .collate = "cols", .to = "tot_var") %>%
  mutate_at(vars(-tot_var), funs(prop = . / tot_var))
  
# Test if the observations come from the same distribution
var_comp_prop %>% 
  summarize(ks_test = ks.test(x = gt_s_prop, gt_ns_prop)$p.value)
  
# Plot density plots for each type of marker
var_comp_prop %>%
  select(gt_s_prop, gt_ns_prop) %>%
  gather(term, prop) %>%
  ggplot(aes(x = prop, fill = term)) +
  geom_density(alpha = 0.75) +
  theme_bw()

```


Examine MxE by looking at the variance of marker effects in each environment


```{r mxe}

# Load the results
load(file.path(result_dir, "S2MET_marker_eff_by_env.RData"))

# Edit to column name
mar_effect_by_env <- mar_effect_by_env %>% 
  rename(environment = `envi ronment`)


```







Determine the proportion of genetic variance of phenotypic stability ($V_G$) that is explained by sensitive/stable markers

<!-- This may not work because there are no replications -->

<!-- ```{r prop.g.fw} -->

<!-- # Number of model fittings -->
<!-- n_iter <- 10 -->


<!-- ## Calculate relationship matrices -->
<!-- # Extract the traits -->
<!-- trts <- unique(S2_MET_marker_eff_pheno_fw_sig$trait) -->
<!-- # Create a results list -->
<!-- var_comp_list <- vector("list", length(trts)) %>% -->
<!--   set_names(trts) -->

<!-- # Iterate over the traits -->
<!-- for (tr in trts) { -->

<!--   # Extract the data to use -->
<!--   pheno_df <- S2_MET_pheno_fw %>%  -->
<!--     filter(trait == tr, stability_term == "b") %>% -->
<!--     select(line_name, estimate, b_std_error) %>%  -->
<!--     distinct() %>% -->
<!--     droplevels() -->

<!--   # Extract the significant markers -->
<!--   sig_mar_df <- S2_MET_marker_eff_pheno_fw_sig %>%  -->
<!--     filter(trait == tr, significance != "not_significant") -->
<!--   # Extract non-sig markers -->
<!--   non_sig_mar_df <- S2_MET_marker_eff_pheno_fw_sig %>%  -->
<!--     filter(trait == tr, significance == "not_significant") -->

<!--   ## Subset marker relationship matrices -->
<!--   sig_mar <- pull(sig_mar_df, marker) -->
<!--   M_s <- M[,sig_mar] -->
<!--   K_s <- A.mat(X = M_s, min.MAF = 0, max.missing = 1) -->

<!--   non_sig_mar <- pull(non_sig_mar_df, marker) -->
<!--   # Sample markers to construct Ks -->
<!--   non_sig_mar_sample <- replicate(n = n_iter, {sort(sample(non_sig_mar, size = length(sig_mar)))}, simplify = FALSE) -->
<!--   M_ns_list <- map(non_sig_mar_sample, ~M[,.]) -->
<!--   K_ns_list <- map(M_ns_list, ~A.mat(X = ., min.MAF = 0, max.missing = 1)) -->

<!--   ## Model frame/matrices -->
<!--   mf <- model.frame(estimate ~ line_name + b_std_error, pheno_df) -->

<!--   y <- model.response(mf) -->

<!--   # Weights -->
<!--   R <- solve(diag(mf$b_std_error^2)) -->

<!--   X <- model.matrix(~ 1, mf) -->

<!--   # Random effect of genotypic main effect -->
<!--   Z_g <- model.matrix(~ -1 + line_name, mf) %>% -->
<!--     `colnames<-`(., colnames(K_s)) -->

<!--   # Iterate over the list of K_ns matrices -->
<!--   var_comp_out <- K_ns_list %>% -->
<!--     map_df(~sommer::mmer(Y = y, X = X, Z = list(g_s = list(Z = Z_g, K = K_s), g_ns = list(Z = Z_g, K = .)), -->
<!--                          R = list(units = R), silent = T) %>% .$var.comp %>% map_df(unname)) -->

<!--   # Add the output to the list -->
<!--   var_comp_list[[tr]] <- parallel_out -->

<!-- } # Close the trait for loop -->




<!-- ``` -->








